### ISSUE TITLE
`install-ptfe.sh` execution sometimes indefinetly stalls after primary node establishes cluster

### ISSUE THREAD

**Issue by shiro105** [hashicorp/terraform-azurerm-terraform-enterprise#13]
# Specs
Kubernetes v1.13
Rook v1.0.0
Ubuntu 16.04 LTS

# Issue
After I execute the module using `terraform apply`, the apply goes through just fine, but upon inspecting the `/var/log/cloud-init-output.log`, `install-ptfe.sh` never finishes its execution. From the following excerpt in that log:

```
ESC[0;94m⚙  Initialize KubernetesESC[0m
+ kubeadm init --ignore-preflight-errors=all --config /opt/replicated/kubeadm.conf --skip-phases '' --skip-token-print
+ tee /tmp/kubeadm-init
[init] Using Kubernetes version: v1.13.5
[preflight] Running pre-flight checks
[preflight] Pulling images required for setting up a Kubernetes cluster
[preflight] This might take a minute or two, depending on the speed of your internet connection
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Activating the kubelet service
[certs] Using certificateDir folder "/etc/kubernetes/pki"
[certs] Generating "ca" certificate and key
[certs] Generating "apiserver" certificate and key
[certs] apiserver serving cert is signed for DNS names [tfe-vrvsikky-primary-0 kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.clus
ter.local tfe-vrvsikky-api.domain.tk tfe-vrvsikky-api.domain.tk] and IPs [10.96.0.1 10.0.0.12]
[certs] Generating "apiserver-kubelet-client" certificate and key
[certs] Generating "front-proxy-ca" certificate and key
[certs] Generating "front-proxy-client" certificate and key
[certs] Generating "etcd/ca" certificate and key
[certs] Generating "etcd/server" certificate and key
[certs] etcd/server serving cert is signed for DNS names [tfe-vrvsikky-primary-0 localhost] and IPs [10.0.0.12 127.0.0.1 ::1]
[certs] Generating "etcd/peer" certificate and key
[certs] etcd/peer serving cert is signed for DNS names [tfe-vrvsikky-primary-0 localhost] and IPs [10.0.0.12 127.0.0.1 ::1]
[certs] Generating "etcd/healthcheck-client" certificate and key
[certs] Generating "apiserver-etcd-client" certificate and key
[certs] Generating "sa" key and public key
[kubeconfig] Using kubeconfig folder "/etc/kubernetes"
[kubeconfig] Writing "admin.conf" kubeconfig file
[kubeconfig] Writing "kubelet.conf" kubeconfig file
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
[kubeconfig] Writing "scheduler.conf" kubeconfig file
[control-plane] Using manifest folder "/etc/kubernetes/manifests"
[control-plane] Creating static Pod manifest for "kube-apiserver"
[control-plane] Creating static Pod manifest for "kube-controller-manager"
[control-plane] Creating static Pod manifest for "kube-scheduler"
[etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
[apiclient] All control plane components are healthy after 37.678342 seconds                                                                                   [uploadconfig] storing the configuration used in ConfigMap "kubeadm-config" in the "kube-system" Namespace
[kubelet] Creating a ConfigMap "kubelet-config-1.13" in namespace kube-system with the configuration for the kubelets in the cluster                           [patchnode] Uploading the CRI Socket information "/var/run/dockershim.sock" to the Node API object "tfe-vrvsikky-primary-0" as an annotation                   [mark-control-plane] Marking the node tfe-vrvsikky-primary-0 as control-plane by adding the label "node-role.kubernetes.io/master=''"]]
[patchnode] Uploading the CRI Socket information "/var/run/dockershim.sock" to the Node API object "tfe-vrvsikky-primary-0" as an annotation
[mark-control-plane] Marking the node tfe-vrvsikky-primary-0 as control-plane by adding the label "node-role.kubernetes.io/master=''"
[mark-control-plane] Marking the node tfe-vrvsikky-primary-0 as control-plane by adding the taints [node-role.kubernetes.io/master:NoSchedule]
[bootstrap-token] Configuring bootstrap tokens, cluster-info ConfigMap, RBAC Roles
[bootstraptoken] configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
[bootstraptoken] configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
[bootstraptoken] configured RBAC rules to allow certificate rotation for all node client certificates in the cluster
[bootstraptoken] creating the "cluster-info" ConfigMap in the "kube-public" namespace
[addons] Applied essential addon: CoreDNS
[addons] Applied essential addon: kube-proxy

Your Kubernetes master has initialized successfully!

To start using your cluster, you need to run the following as a regular user:
                                                                                                                                                                 mkdir -p $HOME/.kube
                                                                                                                                                                   sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
                                                                                                                                                                     sudo chown $(id -u):$(id -g) $HOME/.kube/config

                                                                                                                                                                     You should now deploy a pod network to the cluster.
                                                                                                                                                                     Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
                                                                                                                                                                       https://kubernetes.io/docs/concepts/cluster-administration/addons/

                                                                                                                                                                       You can now join any number of machines by running the following on each node
                                                                                                                                                                       as root:

                                                                                                                                                                         kubeadm join tfe-vrvsikky-api.domain.tk:6443 --token <value withheld> --discovery-token-ca-cert-hash sha256:025988fd9449f9d7fde8271f9c292918d8c57771e
                                                                                                                                                                         65169aeea0662741410bf8e

                                                                                                                                                                         ESC[0;32m✔ Kubernetes Master InitializedESC[0m
                                                                                                                                                                         ESC[0;32mKubernetes masterESC[0m is running at ESC[0;33mhttps://tfe-vrvsikky-api.domain.tk:6443ESC[0m
                                                                                                                                                                         ESC[0;32mKubeDNSESC[0m is running at ESC[0;33mhttps://tfe-vrvsikky-api.domain.tk:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxyESC[0m

                                                                                                                                                                         To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.]]]]]]]]]]


 resources found.
 ESC[0;94m⚙  deploy weave networkESC[0m
 secret/weave-passwd created
 serviceaccount/weave-net created
 clusterrole.rbac.authorization.k8s.io/weave-net created
 clusterrolebinding.rbac.authorization.k8s.io/weave-net created
 role.rbac.authorization.k8s.io/weave-net created
 rolebinding.rbac.authorization.k8s.io/weave-net created
 daemonset.extensions/weave-net created
 ESC[0;32m✔ weave network deployedESC[0m
 ESC[0;94m⚙  remove NoSchedule taint from master nodeESC[0m
 node/tfe-vrvsikky-primary-0 untainted
 ESC[0;32m✔ master taint removedESC[0m
 ESC[0;94m⚙  Await node readyESC[0m
 ESC[0;32m✔ Master Node Ready!ESC[0m

 + kubectl get nodes
 NAME                     STATUS   ROLES    AGE   VERSION
 tfe-vrvsikky-primary-0   Ready    master   31s   v1.13.5
 ESC[0;32m✔ Kubernetes nodesESC[0m

 + kubectl get pods -n kube-system
 NAME                       READY   STATUS              RESTARTS   AGE
 coredns-86c58d9df4-dn7vv   0/1     ContainerCreating   0          22s
 coredns-86c58d9df4-jvwnb   0/1     ContainerCreating   0          22s
 kube-proxy-5fg55           1/1     Running             0          21s
 weave-net-tv9sp            2/2     Running             0          21s                                                                                          ESC[0;32m✔ Kubernetes systemESC[0m

 ESC[0;94m⚙  deploy rookESC[0m
 namespace/rook-ceph-system created
 customresourcedefinition.apiextensions.k8s.io/cephclusters.ceph.rook.io created
 customresourcedefinition.apiextensions.k8s.io/cephfilesystems.ceph.rook.io created
 customresourcedefinition.apiextensions.k8s.io/cephnfses.ceph.rook.io created
 customresourcedefinition.apiextensions.k8s.io/cephobjectstores.ceph.rook.io created
 customresourcedefinition.apiextensions.k8s.io/cephobjectstoreusers.ceph.rook.io created
 customresourcedefinition.apiextensions.k8s.io/cephblockpools.ceph.rook.io created
 customresourcedefinition.apiextensions.k8s.io/volumes.rook.io created
 clusterrole.rbac.authorization.k8s.io/rook-ceph-cluster-mgmt created
 role.rbac.authorization.k8s.io/rook-ceph-system created
 clusterrole.rbac.authorization.k8s.io/rook-ceph-global created
 clusterrole.rbac.authorization.k8s.io/rook-ceph-mgr-cluster created
 serviceaccount/rook-ceph-system created]]]]]]]]]]]]]]]]]]
 clusterrolebinding.rbac.authorization.k8s.io/rook-ceph-global created
 deployment.apps/rook-ceph-operator created
 ESC[0;94m⚙  Await rook readyESC[0m
 ESC[0;32m✔ Rook Ready!ESC[0m
 namespace/rook-ceph created
 serviceaccount/rook-ceph-osd created
 serviceaccount/rook-ceph-mgr created
 role.rbac.authorization.k8s.io/rook-ceph-osd created
 clusterrole.rbac.authorization.k8s.io/rook-ceph-mgr-system created
 role.rbac.authorization.k8s.io/rook-ceph-mgr created
 rolebinding.rbac.authorization.k8s.io/rook-ceph-cluster-mgmt created
 rolebinding.rbac.authorization.k8s.io/rook-ceph-osd created
 rolebinding.rbac.authorization.k8s.io/rook-ceph-mgr created
 rolebinding.rbac.authorization.k8s.io/rook-ceph-mgr-system created
 clusterrolebinding.rbac.authorization.k8s.io/rook-ceph-mgr-cluster created
 cephcluster.ceph.rook.io/rook-ceph created
 cephblockpool.ceph.rook.io/replicapool created
 storageclass.storage.k8s.io/default created
 [|]
```

once the script finished creating the Rook and creates next batch of Kubernetes objects, the script just stops at that point. I tried looking at the Kubernetes objects that it made, and I do not see anything in particular that looks off:

```
$ kubectl get pods --all-namespaces
NAMESPACE          NAME                                             READY   STATUS    RESTARTS   AGE
kube-system        coredns-86c58d9df4-dn7vv                         1/1     Running   0          131m
kube-system        coredns-86c58d9df4-jvwnb                         1/1     Running   0          131m
kube-system        etcd-tfe-vrvsikky-primary-0                      1/1     Running   0          130m
kube-system        kube-apiserver-tfe-vrvsikky-primary-0            1/1     Running   0          130m
kube-system        kube-controller-manager-tfe-vrvsikky-primary-0   1/1     Running   2          130m
kube-system        kube-proxy-5fg55                                 1/1     Running   0          131m
kube-system        kube-scheduler-tfe-vrvsikky-primary-0            1/1     Running   2          130m
kube-system        weave-net-tv9sp                                  2/2     Running   0          131m
rook-ceph-system   rook-ceph-agent-hnm2l                            1/1     Running   0          129m
rook-ceph-system   rook-ceph-operator-59bb564fcf-vvz4n              1/1     Running   1          130m
rook-ceph-system   rook-discover-4z6f8                              1/1     Running   0          129m
```

```
$ kubectl get deployments --all-namespaces
NAMESPACE          NAME                 READY   UP-TO-DATE   AVAILABLE   AGE
kube-system        coredns              2/2     2            2           133m
rook-ceph-system   rook-ceph-operator   1/1     1            1           133m
```

Upon checking the logs for the `kubelet`, I did find multiple lines of the following right before the kubelet did not produce any more logs:

```
$ journalctl -u kubelet
driver-call.go:267] Failed to unmarshal output for command: init, output: "", error: unexpected end of JSON input
driver-call.go:150] FlexVolume: driver call failed: executable: /usr/libexec/kubernetes/kubelet-plugins/volume/exec/ceph.rook.io~rook-ceph-system/rook-ceph-system, args: [init], error: fork/exec /usr/libexec/kubernetes/kubelet-plugins/volume/exec/ceph.rook.io~rook-ceph-system/rook-ceph-system: no such file or directory, output: ""
```

# Outcome 
The cluster remains only having the primary node in it, with all of the other primary and secondaries indefinetly waiting on the primary to finish setting up the pre requirements for the cluster. As such, accessing the Dashboard will not work, nor will accessing the health check produce anything. 

The most puzzling aspect of this is that this behavior only happens _50%_ of the time I run this module. If I were to do a `terraform destroy` on the module and rebuild it using `terraform apply`, the application will successfully spin up. What's more, the same logs that I showed for the `kubelet` will _still_ appear. The only difference is that the installation will continue past the final logs that I showed for the `install-ptfe.sh`

# Outside References
Upon doing some investigation on Rook and Ceph, I did stumble across [this issue](https://github.com/rook/rook/issues/1501) again. They seemed to have fixed it in [this PR](https://github.com/rook/rook/pull/2245), but if what I am seeing here is true, then that previous issue may have returned again.

**Comment by shiro105** [comment]
Just wanted to circle back on this issue. I have not ran into this error for some time ever since I documented this issue as well as our support ticket call a few months back. As such, I don't believe it is a problem anymore, since this may have been patched in an update after this issue has been documented. As such, I will close this issue. If it comes up again, I'll open up a new ticket.
